iceoryx_posh/test/moduletests/test_popo_server_port_roudi.cpp

```diff
@@ -0,0 +1,262 @@
+    auto& sut = serverPortWithOfferOnCreate;
+
+    sut.portUser.offer();
+
+    sut.portRouDi.tryGetCaProMessage()
+        .and_then(
+            [&](const auto& caproMessage) { FAIL() << "Expected no CaPro message but got: " << caproMessage.m_type; })
+        .or_else([&]() { SUCCEED(); });
+}
+
+TEST_F(ServerPort_test, TryGetCaProMessageOnStopOfferWhenOfferingHasCaProMessageTypeStopOffer)
+{
+    ::testing::Test::RecordProperty("TEST_ID", "83467e90-734b-4e51-836c-2dbeaf44ce95");
+    auto& sut = serverPortWithOfferOnCreate;
+
+    sut.portUser.stopOffer();
+
+    sut.portRouDi.tryGetCaProMessage()
+        .and_then([&](const auto& caproMessage) { EXPECT_THAT(caproMessage.m_type, Eq(CaproMessageType::STOP_OFFER)); })
+        .or_else([&]() { FAIL() << "Expected CaPro message but got none"; });
+}
+
+TEST_F(ServerPort_test, TryGetCaProMessageOnStopOfferWhenNotOfferingHasNullopt)
+{
+    ::testing::Test::RecordProperty("TEST_ID", "a9a162d3-307a-4add-af23-63511da4b07e");
+    auto& sut = serverPortWithoutOfferOnCreate;
+
+    sut.portUser.stopOffer();
+
+    sut.portRouDi.tryGetCaProMessage()
+        .and_then(
+            [&](const auto& caproMessage) { FAIL() << "Expected no CaPro message but got: " << caproMessage.m_type; })
+        .or_else([&]() { SUCCEED(); });
+}
+
+// END tryGetCaProMessage tests
+
+// BEGIN test CaPro transitions
+
+TEST_F(ServerPort_test, StateNotOfferedWithCaProMessageTypeConnectReactsWithNack)
+{
+    ::testing::Test::RecordProperty("TEST_ID", "ceaef856-2a8d-46c0-9167-fe1ca6fad736");
+    auto& sut = serverPortWithoutOfferOnCreate;
+
+    auto caproMessage = CaproMessage{CaproMessageType::CONNECT, sut.portData.m_serviceDescription};
+
+    sut.portRouDi.dispatchCaProMessageAndGetPossibleResponse(caproMessage)
+        .and_then([&](const auto& responseCaproMessage) {
+            EXPECT_THAT(responseCaproMessage.m_serviceDescription, Eq(sut.portData.m_serviceDescription));
+            EXPECT_THAT(responseCaproMessage.m_type, Eq(iox::capro::CaproMessageType::NACK));
+        })
+        .or_else([&]() { FAIL() << "Expected CaPro message but got none"; });
+}
+
+TEST_F(ServerPort_test, StateNotOfferedWithCaProMessageTypeDisconnectReactsWithNack)
+{
+    ::testing::Test::RecordProperty("TEST_ID", "0bde368a-b2a6-4c73-b668-43afda427f56");
+    auto& sut = serverPortWithoutOfferOnCreate;
+
+    auto caproMessage = CaproMessage{CaproMessageType::DISCONNECT, sut.portData.m_serviceDescription};
+
+    sut.portRouDi.dispatchCaProMessageAndGetPossibleResponse(caproMessage)
+        .and_then([&](const auto& responseCaproMessage) {
+            EXPECT_THAT(responseCaproMessage.m_serviceDescription, Eq(sut.portData.m_serviceDescription));
+            EXPECT_THAT(responseCaproMessage.m_type, Eq(iox::capro::CaproMessageType::NACK));
+        })
+        .or_else([&]() { FAIL() << "Expected CaPro message but got none"; });
+}
+
+TEST_F(ServerPort_test, StateNotOfferedWithCaProMessageTypeStopOfferReactsWithNack)
```

Comment:
I think you miss a test for a capro protocol violation. Could you add this for state not offered and state offered.
Maybe you could send `CaproMessageType::SUB` or `CaproMessageType::PING` and check for the violation.